examples/remote/binary_dependencies/cargo/crates.bzl

```diff
@@ -9,6 +9,144 @@ load("@bazel_tools//tools/build_defs/repo:git.bzl", "new_git_repository")  # bui
+            crates.append(dependencies[native.package_name()][crate])
+
+    if errors:
+        fail("Missing crates `{}` for package `{}`. Available crates `{}".format(
+            errors,
+            native.package_name(),
+            dependencies[native.package_name()],
+        ))
+
+    return crates
+
+def all_crates(normal = False, proc_macro = False, dev = False, dev_only = False):
+    """EXPERIMENTAL -- MAY CHANGE AT ANY TIME: Finds the fully qualified label of all requested direct crate dependencies \
+    for the package where this macro is called.
+
+    If no parameters are set, all normal and proc_macro dependencies are returned.
+    Setting any one flag will otherwise impact the contents of the returned list
+
+    WARNING: This macro is part of an expeirmental API and is subject to change.
+
+    Args:
+        normal (bool, optional): If True, normal dependencies are included in the
+            output list. Defaults to False.
+        proc_macro (bool, optional): If True, proc_macro dependencies will be
+            included in the output list. Defaults to False.
+        dev (bool, optional): If True, dev dependencies are included when the
+            `normal` and `proc_macro` parameters are used. Defaults to False.
+        dev_only (bool, optional): If True, only development dependencies will be
+            returned by this list. This paramter otherwise follows the same rules
+            as `dev`. Defaults to False.
+
+    Returns:
+        list: A list of labels to cargo-raze generated targets (str)
+    """
+
+    # Determine the relevant maps to use
+    dependencies = dict()
+    all_maps = []
+    if normal:
+        if not dev_only:
+            all_maps.append(_DEPENDENCIES)
+        if dev or dev_only:
+            all_maps.append(_DEV_DEPENDENCIES)
+    if proc_macro:
+        if not dev_only:
+            all_maps.append(_PROC_MACRO_DEPENDENCIES)
+        if dev or dev_only:
+            all_maps.append(_DEV_PROC_MACRO_DEPENDENCIES)
+
+    # Default to always using normal dependencies
+    if not all_maps:
+        if not dev_only:
+            all_maps.append(_DEPENDENCIES)
+        if dev or dev_only:
+            all_maps.append(_DEV_DEPENDENCIES)
+
+    if not all_maps:
+        fail("Failed to add at least 1 map to the `all_maps` list with parameters: " +
+             "normal = {normal}, proc_macro = {proc_macro}, dev = {dev}, dev_only = {dev_only}".format(
+                 normal = normal,
+                 proc_macro = proc_macro,
+                 dev = dev,
+                 dev_only = dev_only,
+             ))
+
+    for dep_map in all_maps:
+        for package_name in dep_map:
+            if package_name in dependencies:
+                dependencies[package_name].extend(dep_map[package_name])
+            else:
+                dependencies[package_name] = dep_map[package_name]
+
+    if not dependencies:
+        return []
+
+    return dependencies[native.package_name()].values()
```

Comment:
I feel less strongly about this one, but here too can we take package_name as an argument?

Reply:
It can be an optional argument. But I don't want to force users to have to write that.

Reply:
Why not?

Reply:
This goes back to my desire to not have to write things unnecessarily. Why *would* we want to force users to always specify a package name? This to me is the purpose of `native.package_name()`. Macros can understand what package they're being invoked in. The only use case for providing an explicit package name is if you were trying to use dependencies from another package.

Reply:
I am anti magic, and this is magic.

I suggest one compromise above to break this into two macros: one "non magical" version, and one "magical" version. The magical version communciates that the output is dependent on the current package ("here").

Reply:
That works for me. What should these macros be called then? Anything better than `*_here`? Not the worst, just wondering if there's anything better.

Reply:
What do you think about 

`resolved_crate_deps(package_name = native.package_name(), crate_names)

(overload the same function). You get the magic for free, but the explicit version is also available. If mandatory args can't follow optional args, then 

`resolved_crate_deps(crate_names, package_name = native.package_name())` is an an option.

Reply:
I became more open to this overload because I did a github wide search for native.package_name and realized what this is actually doing -- it's just returning the name of the build rule itself. That's less serious than what I thought it was, which was a fully qualified build target.

Reply:
Can we go with `crate_deps`? Can `resolved` be implied?

Reply:
I prefer the verbose name, but not strongly enough to effort an argument or continue pushing back.

Reply:
I care a lot about this name being intuitive and not obnoxious but I value the correctness of it as well. If I can get your thoughts on https://github.com/google/cargo-raze/pull/319#discussion_r548377332 I can change it.

Reply:
Thanks for responding there 🙏